src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -292,8 +319,131 @@ internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort a
+            Compilation compilation,
+            SyntaxNode syntaxNode,
+            DiagnosticBag diagnostics)
+        {
+            return _dataSectionStringLiteralTypes.GetOrAdd(text, static (key, arg) =>
+            {
+                var (@this, data, compilation, syntaxNode, diagnostics) = arg;
+
+                string name = "<S>" + DataToHexViaXxHash128(data);
+
+                MappedField dataField = @this.CreateDataField(data, alignment: 1);
+
+                Cci.IMethodDefinition bytesToStringHelper = @this.GetOrSynthesizeBytesToStringHelper(compilation, syntaxNode, diagnostics);
+
+                return new DataSectionStringType(
+                    module: (ITokenDeferral)@this._moduleBuilder,
+                    name: name,
+                    systemObject: @this._systemObject,
+                    systemString: @this._systemStringType,
+                    containingType: @this,
+                    dataField: dataField,
+                    bytesToStringHelper: bytesToStringHelper,
+                    diagnostics: diagnostics);
+            },
+            (this, data, compilation, syntaxNode, diagnostics));
+        }
+
+        /// <summary>
+        /// Gets the <see cref="BytesToStringHelper"/> or creates it if it does not exist yet.
+        /// </summary>
+        private Cci.IMethodDefinition GetOrSynthesizeBytesToStringHelper(Compilation compilation, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        {
+            var method = GetMethod(SynthesizedBytesToStringFunctionName);
+
+            if (method is null)
+            {
+                TryAddSynthesizedMethod(synthesizeBytesToStringHelper(this, compilation, syntaxNode, diagnostics));
+
+                method = GetMethod(SynthesizedBytesToStringFunctionName);
+            }
+
+            Debug.Assert(method is not null);
+            return method;
+
+            static Cci.IMethodDefinition synthesizeBytesToStringHelper(
+                PrivateImplementationDetails @this,
+                Compilation compilation,
+                SyntaxNode syntaxNode,
+                DiagnosticBag diagnostics)
+            {
+                if (!tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__get_UTF8, out var encodingUtf8) |
+                    !tryGetWellKnownTypeMember(compilation, syntaxNode, diagnostics, WellKnownMember.System_Text_Encoding__GetString, out var encodingGetString))
+                {
+                    return new DefaultMethodDef(
+                        name: SynthesizedBytesToStringFunctionName,
+                        visibility: Cci.TypeMemberVisibility.Private,
+                        containingType: @this,
+                        maxStack: 8,
+                        il: []);
+                }
+
+                Debug.Assert(encodingUtf8 is not null && encodingGetString is not null);
+
+                var ilBuilder = new ILBuilder(
```

Comment:
>var ilBuilder = new ILBuilder( [](http://example.com/codeflow?start=16&length=30)

I think it would be better to encapsulate the IL emit logic in ```BytesToStringHelper``` type.